docs/man/ufod.1.txt
ufod(1)
=======

NAME
----
ufod - Start ufo daemon


SYNOPSIS
--------
[verse]
'ufo-query' [-l <address>]


DESCRIPTION
-----------

Start the ufod daemon for connection requests by a master process.


OPTIONS
-------

*--listen*::
*-l*::
        Interface and port to listen for incoming requests.


NEWS
=========
Changelog
=========

Here you can see the full list of changes between each ufo-core release.


Version 0.11.1
==============

Bugfix release released on November 12th 2016.

- Install systemd unit file through pkg-config
- Fix #110: install templates correctly
- Do not run xmllint on manpage generation output
- Remove unused CMake modules
- Remove executable bit from source files
- Remove PACKAGE_* variables


Version 0.11
============

Released on November 8th 2016

Enhancements:

- Build manpages for the tools
- Update TomoPy integration
- Add UFO_DEVICE_TYPE environment variable
- Properly build on MacOS
- Unify debug message output format
- ufo-launch: rewrite specification parser allowing more flexible descriptions
- ufo-launch: add --quieter
- ufo-mkfilter: add --type and --use-gpu

Fixes:

- Check if we have with multiple roots to exit early
- Show version information consistently
- ufo-prof: support Python < 3.0

Breaks:

- GNUInstallDirs instead is now used instead of our own ConfigurePaths CMake
  module which might affect installation paths on your system. Please note that
  for example -DPREFIX thus becomes -DCMAKE_INSTALL_PREFIX.
- Remove CPack
- ufod: unused --paths option removed
- ufo-runjson: unused --path option removed
- ufo-launch: do not execute graph when --dump'ing


Version 0.10
============

Released on May 24th 2016

Enhancements:

- Add UFO_GPU_NODE_INFO_MAX_MEM_ALLOC_SIZE
- Fix #103: allow ufo-launch to use arbitrary graphs with new workflow
  specification and more robust parsing
- Look for AMD APP SDK 3.0
- Cache programs to avoid rebuilding them
- ufo-runjson now also outputs the number of processed items
- Fix #104: output type and blurb with -v

Fixes:

- Fix #106: match word characters to find plugins
- Fix #101: really unref non-intermediate nodes
- ufo-launch: query only valid graph
- Fixed misleading documentation
- Fixed some leaks and unreferenced resources
- Stop num-processed from being serialized into JSON field
- ufod: fix segfault if no address is specified

Breaks:

- ufo-query: print errors on stderr instead of stdout
- ufo-launch: removed --progress and --time in favor of --quiet
- Removed public ufo_signal_emit symbol
- Use date and time to differentiate trace profiles
- Replace clprof with ufo-prof


Version 0.9.1
=============

Release on January 12th 2016.

Enhancements:

- Pass a property map to scheduler in the Python wrapper
- Added convenience wrapper to copy into a buffer

Fixes:

- Fix version numbering from 0.8.x to 0.9.x
- Fix documentation issues
- Fix build problems if libzmq is not present
- ufo-launch unrefs non-intermediate nodes


Version 0.9
===========

Release on November 3rd 2015.

Enhancements:

- Generally improved debug output
- Add support for plugin packages as demonstrated by the ART plugins
- Add Docker installation method
- Add simple fabfile to start ufod instances
- Allow reductors to pause processing
- Add ufo-query binary to retrieve info about tasks
- Add ufo_buffer_set_device_array API call
- Python: add ufo.numpy.empty_like
- ufo-launch: add --dump flag to serialize to JSON
- ufo-launch: add Bash completion script
- ufo-launch: pass address list

Fixes:

- Fix segfault with long-running tasks
- Prevent daemon from leaking OpenCL resources
- Fix broken continuous daemon operation
- ufo-launch: fix parsing uint64
- Use same reductor policy for both schedulers
- Fix ufo_buffer_new_with_data
- Fix nano second to second conversion

Breaks:

- Specify device subset with UFO_DEVICES
- Do not prepend . to trace and profile output
- Change ufo_buffer_set_host_array signature
- Use common timestamp unit for both trace types


Version 0.8
===========

Release on May 19th 2015.

This release breaks with the distinction of ArchGraph and Resources. The former
is removed with its functionality moved to the latter. Besides that numerous
improvements have been incorporated:

- ufo-launch: add --time and --trace options
- Added a systemd unit for ufod
- Fixed cluster communication.
- Fixed #86: use CL_INTENSITY instead of CL_R
- Handle kernel path with environment variable UFO_KERNEL_PATH
- ufo-launch now parses boolean properties
- Removed hard dependency on ZeroMQ
- Removed WITH_DEBUG and WITH_PROFILING CMake variables
- Add JSON-based TANGO server for remote computation
- Removed ufo_base_scheduler_get_gpu_nodes()
- Fixed #80: add ufo_gpu_node_get_info()
- Fixed re-running the same task graph
- Removed tifffile dependency
- Fixed #78: reset num-processed before execution
- Fixed #77: initialize threads when accessing the GIL


Version 0.7
===========

Released on February 20th 2015.

This is a major update from the previous version released a year ago. Besides
numerous bug fixes and compatibility enhancements we

- added ufo_signal_emit for threadsafe signal emission
- export plugin and kernel directories via pkg-config so that thirdparty plugins
  easily know where to install themselves
- added the ufo-launch tool to run basic pipelines directly from the command
  line
- merged the ufo-python-tools and ufo-tests repos with the core repository
- added additional buffer depths
- parse the UFO_PLUGIN_PATH environment variable to specify additional plugin
  locations
- added a copy task which copies data from input 0
- added a fixed scheduler for manual assignment of hardware resources to task
  nodes
- added "processed" and "generated" signals to tasks which emit whenever either
  action completed
- added UFO_DEVICE_TYPE_ACC for accelerator devices
- added buffer views for larger-than-GPU data
- output OpenCL profiler information as Chrome JSON
- added ufo_resources_get_kernel_with_opts()
- added ufo_buffer_set_host_array()
- added ufo_buffer_get/set_metadata(), ufo_buffer_get_metadata_keys() and
  ufo_buffer_copy_meta_data()
- added ufo_buffer_get_device_array_with_offset()
- added ufo_buffer_get_location()

and broke compatibility by

- retiring the UfoConfig infrastructure and

tools/ufod.c
/*
 * Copyright (C) 2011-2013 Karlsruhe Institute of Technology
 *
 * This file is part of ufod.
 *
 * ufod is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation, either
 * version 3 of the License, or (at your option) any later version.
 *
 * ufod is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with ufod.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"

#ifdef __APPLE__
#include <OpenCL/cl.h>
#else
#include <CL/cl.h>
#endif
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#include <ufo/ufo.h>

static UfoDaemon *global_daemon;

typedef struct {
    gchar **paths;
    gchar *addr;
} Options;

static Options *
opts_parse (gint *argc, gchar ***argv)
{
    GOptionContext *context;
    Options *opts;
    gboolean show_version = FALSE;
    GError *error = NULL;

    opts = g_new0 (Options, 1);

    GOptionEntry entries[] = {
        { "listen", 'l', 0, G_OPTION_ARG_STRING, &opts->addr,
          "Address to listen on (see http://api.zeromq.org/3-2:zmq-tcp)", NULL },
        { "path", 'p', 0, G_OPTION_ARG_STRING_ARRAY, &opts->paths,
          "Path to node plugins or OpenCL kernels", NULL },
        { "version", 'v', 0, G_OPTION_ARG_NONE, &show_version,
          "Show version information", NULL },
        { NULL }
    };

    context = g_option_context_new ("FILE");
    g_option_context_add_main_entries (context, entries, NULL);

    if (!g_option_context_parse (context, argc, argv, &error)) {
        g_print ("Option parsing failed: %s\n", error->message);
        g_free (opts);
        return NULL;
    }

    if (show_version) {
        g_print ("ufod %s\n", UFO_VERSION);
        exit (EXIT_SUCCESS);
    }

    if (opts->addr == NULL)
        opts->addr = g_strdup ("tcp://*:5555");

    g_option_context_free (context);

    return opts;
}

static void
opts_free (Options *opts)
{
    g_strfreev (opts->paths);
    g_free (opts->addr);
    g_free (opts);
}

static void
terminate (int signum)
{
    GError *error = NULL;

    if (signum == SIGTERM)
        g_print ("Received SIGTERM, exiting...\n");

    if (signum == SIGINT)
        g_print ("Received SIGINT, exiting...\n");

    if (global_daemon != NULL) {
        ufo_daemon_stop (global_daemon, &error);

        if (error != NULL)
            g_printerr ("Error: %s\n", error->message);

        g_object_unref (global_daemon);
    }

    exit (EXIT_SUCCESS);
}

int
main (int argc, char * argv[])
{
    Options *opts;
    GError *error = NULL;

    g_type_init ();
    g_thread_init (NULL);

    if ((opts = opts_parse (&argc, &argv)) == NULL)
        return 1;

    /* Now, install SIGTERM/SIGINT handler */
    (void) signal (SIGTERM, terminate);
    (void) signal (SIGINT, terminate);

    global_daemon = ufo_daemon_new (opts->addr);
    ufo_daemon_start (global_daemon, &error);

    if (error != NULL) {
        g_printerr ("Error: %s\n", error->message);
        g_object_unref (global_daemon);
        return 1;
    }

    g_print ("ufod %s - waiting for requests on %s ...\n", UFO_VERSION,
                                                           opts->addr);

    while (TRUE) {
        g_usleep (G_MAXULONG);
    }

    opts_free (opts);
    return 0;
}
